xline/src/request_validation.rs

```diff
@@ -177,36 +172,106 @@ impl RequestValidator for AuthRoleAddRequest {
 impl RequestValidator for AuthRoleGrantPermissionRequest {
     fn validation(&self) -> Result<(), ValidationError> {
         if self.perm.is_none() {
-            return Err(ValidationError::new("Permission not given"));
+            return Err(ValidationError::PermissionNotGiven);
         }
 
         Ok(())
     }
 }
 
 /// Error type in Validation
-#[derive(Error, Debug)]
-#[error("{0}")]
-pub struct ValidationError(String);
-
-impl ValidationError {
-    /// Creates a new `ValidationError`
-    fn new(message: impl Into<String>) -> Self {
-        Self(message.into())
-    }
+#[derive(Error, Debug, Copy, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub enum ValidationError {
+    /// Key is not provided
+    #[error("key is not provided")]
+    EmptyKey,
+    /// Ignore value is set but value is provided
+    #[error("ignore value is set but value is provided")]
+    ValueProvided,
+    /// Ignore lease is set but lease is provided
+    #[error("ignore lease is set but lease is provided")]
+    LeaseProvided,
+    /// Invalid sort option
+    #[error("invalid sort option")]
+    InvalidSortOption,
+    /// Too many operations in txn request
+    #[error("too many operations in txn request")]
+    TooManyOps,
+    /// Request not provided in operation
+    #[error("request not provided in operation")]
+    RequestNotProvided,
+    /// Duplicate key given in txn request
+    #[error("duplicate key given in txn request")]
+    DuplicateKey,
+    /// User name is empty
+    #[error("user name is empty")]
+    UserEmpty,
+    /// Password is empty
+    #[error("password is empty")]
+    PasswordEmpty,
+    /// Role name is empty
+    #[error("role name is empty")]
+    RoleEmpty,
+    /// Permission not given
+    #[error("permission not given")]
+    PermissionNotGiven,
 }
 
+// As etcd client use grpc error message to parse error types, this is to make the error
```

Comment:
we can link the relevant code here to prove our opinion. It should include the error definitions(error.go) and how etcd client uses these errors (client/doc.go)

Reply:
done